### PR TITLE
feat: separate vaults launch from platform

### LINKED ISSUE #5819: install Inter protocol defi contracts after boostrap using publishBundle
## Problem Statement?

Unlike the PSM contract, which can be considered part of the platform, vaults introduce defi considerations that motivate launching them as a separate governance action.

## Description of the Design

Somewhat like the RUN Protocol Preview (#5062), but use `publishBundle` (#4975) and hash bundles to avoid putting the full text of the bundles inside swingset messages.

## Documentation Considerations

A runbook seems in order, much like ...

 - [Runbook: RUN Protocol Preview · Agoric/agoric\-sdk Wiki](https://github.com/Agoric/agoric-sdk/wiki/Runbook:-RUN-Protocol-Preview)

See also:
 - https://github.com/Agoric/agoric-sdk/issues/7493

## Security Considerations

see also

 - #4280
 - #4395

## Test Plan

Note that the testnets/devnets we typically stand up lately using instagoric avoid this problem by using a swingset config (derived from `decentral-demo-config.json`) where the contracts get installed as part of bootstrap. So testing this involves

 1. start a chain _without_ the Inter protocol contracts
 2. Do enough governance to get the Vaults working
    1. generate proposals to install them (`agoric run inter-protocol/scripts/init-core.js`)
        - including any `add-collateral-core.js` proposals to add IBC assets to the VBANK and such. 
    2. submit the proposals using `agd gov submit-proposal swingset-core-eval ...` as in `cosmic-swingset/Makefile`
    3. vote on the proposals
    4. Exercise gov committee, Oracle invitations to set up the Vaults, Auction, etc.
 3. Verify that you can open a vault, get liquidated, etc.

optional goals:
   - static typing for coreProposals

cc @warner

### PR DESCRIPTION
**Pull request by dckc**
closes: #5819 (or at least: refs)

_The code still needs some rebasing / fixup, but I works._

@warner do you want to review this? feel free to take yourself out of the list.

## Description

 - [x] remove vaults proposal from devnet bootstrap config
 - [x] take governance committee, oracles, out of vaults proposal
 - [x] take stakeFactory out of vaults proposal
 - [x] add initial collateral type
 - [ ] prune dead code from `packages/inter-protocol/scripts/init-core.js`, `packages/inter-protocol/src/proposals/core-proposal.js`

 - [ ] review #5819 etc. for configuration details to take care of

### Security Considerations

nothing new, I don't think

### Scaling Considerations

 - [x] verify that the largest crank is smaller than the block limit, suggesting that work would be broken up between blocks expectedly

### Documentation Considerations

The proposal to launch vaults should come with a community discussion post.

See also
 - #7493 

### Testing Considerations

 - [x] bootstrap test
 - [x] test that initial collateral type is installed, by checking the metrics for list of collaterals
 
TODO:
 - [ ] wire oracles
 - [ ] test that vaults are fully functional by opening a vault?
 - [ ] test that we start econCommitteeCharter only once (https://github.com/Agoric/agoric-sdk/pull/7426#discussion_r1168855675)
 - [ ] test enacting the proposal in a live chain; in particular

### CODE REVIEW

File: packages/deploy-script-support/src/cachedBundleSpec.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    const cacheFile = pathResolve(cacheDir, `${bundleID}.json`);
+
+    await fs.mkdir(cacheDir, { recursive: true });
+    try {
+      await fs.stat(cacheFile);
+    } catch (e) {
+      if (e.code !== 'ENOENT') {
+        throw e;
+      }
+      // FIXME: We could take a hard dependency on `tmp` here, but is it worth it?
```
Comment (turadg) [inline packages/deploy-script-support/src/cachedBundleSpec.js]: probably not worth the NPM dependency but we do have Node and could use [mkdtemp](https://nodejs.org/api/fs.html#fspromisesmkdtempprefix-options)

File: packages/inter-protocol/src/proposals/econ-behaviors.js [hunk 2]
```diff
@@ -355,7 +355,7 @@ export const startRewardDistributor = async ({
     bankManager,
     vaultFactoryKit,
     periodicFeeCollectors,
-    stakeFactoryKit,
+    // stakeFactoryKit,
```
Comment (turadg) [inline packages/inter-protocol/src/proposals/econ-behaviors.js]: simply delete for now

File: packages/inter-protocol/src/proposals/econ-behaviors.js [hunk 3]
```diff
@@ -765,3 +765,44 @@ export const startStakeFactory = async (
   ]);
 };
 harden(startStakeFactory);
+
+/**
+ * Introduce charter to governed creator facets.
+ *
+ * XXX make this the responsibility of startVaultFactory, startReserve, etc.?
```
Comment (turadg) [inline packages/inter-protocol/src/proposals/econ-behaviors.js:796]: I think so. why not?

File: packages/vats/test/bootstrapTests/test-inter-launch.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+  const bundles = [];
+  for (const mm of txt.matchAll(/swingset install-bundle @([^\n]+)/gm)) {
+    const [_, bundle] = mm;
+    bundles.push(bundle);
+  }
+  return { permit, script, bundles };
+};
+
+test.serial('make vaults launch proposal', async t => {
```
Comment (turadg) [inline packages/vats/test/bootstrapTests/test-inter-launch.js:186]: yay bootstrap test!

**Review by turadg — COMMENTED** [review]
A first pass. The testing plan LGTM. Might break loadgen? if so you might want to force integration after getting this synced again with master

File: packages/deploy-script-support/src/cachedBundleSpec.js (hunk 1 above)
Comment (dckc) [reply packages/deploy-script-support/src/cachedBundleSpec.js]: that file isn't actually in this PR; it's an artifact of stacking this on `mfig-cli-run-no-solo`

File: packages/inter-protocol/src/proposals/econ-behaviors.js (hunk 3 above)
Comment (dckc) [reply packages/inter-protocol/src/proposals/econ-behaviors.js:796]: just time.

File: packages/vats/src/core/lib-boot.js [hunk 5]
```diff
@@ -171,13 +171,13 @@ export const makeBootstrap = (
       const { root } = vat;
       const decodedArgs = args.map(decodePassable);
       const result = await E(root)[methodName](...decodedArgs);
-      return encodePassable(result);
+      return encodePassable(harden(result));
```
Comment (michaelfig) [inline packages/vats/src/core/lib-boot.js:174]: Good catch!

**Review by michaelfig — COMMENTED** [review]
Looking good so far!

**Comment by mhofman** [comment]
Updated the PR with some cleanups around proposal builder, installing the ATOM collateral, measuring per crank computrons. All tests pass, but the PR still needs some rebasing / fixups as I mostly kept the original WIP commits.

**Comment by dckc** [comment]
@mhofman thanks for polishing this up, but as noted in https://github.com/Agoric/agoric-sdk/issues/5819#issuecomment-1533863090 , this is no longer the plan.

If you added anything along the way that should be kept, please put it in another PR.

**Pull request closed without merging**